<div dir="rtl" style="text-align:right">
<h1 id="دفتر-المحاضرة-الكامل-2">دفتر المحاضرة الكامل — 2</h1>
<p dir="rtl" style="text-align:right">نقرأ الشرح، ونتوقّع النتيجة، ثم ننفّذ الخلية التالية ونفسّر مخرجاتها قبل أن نتابع. نستخدم <strong>Restart Kernel and Run All</strong> لنتأكد من أن العمل لا يعتمد على جلسة سابقة. تبقى الشروح والاشتقاقات الرياضية والافتراضات الفيزيائية من الفصل محيطة بالخلايا.</p>
</div>

<div dir="rtl" style="text-align:right">
<h2 id="تهيئة-الدفتر-المرفقة">تهيئة الدفتر المرفقة</h2>
<p dir="rtl" style="text-align:right">نختار نواة <strong>Julia 1.12</strong> ونفتح الدفتر من مجلد فصله. تعثر هذه التهيئة المرفقة على بيئة المقرر المسجّلة وتنشئ مجلدًا منفصلًا للأشكال المصدّرة؛ وهي لا تثبّت أي حزم. أوامر التهيئة بنية تحتية، وليست مادة لغوية جديدة تخضع للتقييم.</p>
</div>

In [1]:
import Pkg
course_root = abspath("../..")
if !isfile(joinpath(course_root, "Project.toml"))
    error("Start this notebook from its week folder, e.g. week-1/notebooks")
end
Pkg.activate(course_root; io=devnull)
output_dir = get(ENV, "COURSE_NOTEBOOK_OUTPUT", joinpath(pwd(), "generated", "lecture-ar"))
mkpath(output_dir)
println("Course environment active; figure output directory ready.")

Course environment active; figure output directory ready.


<div dir="rtl" style="text-align:right">
<h1 id="الأسبوع-2-القيم-والقرارات-وجسم-على-منحدر">الأسبوع 2: القيم والقرارات وجسم على منحدر</h1>
<h2 id="محاضرة-من-المتغيرات-إلى-الخوارزمية">محاضرة: من المتغيرات إلى الخوارزمية</h2>
<p dir="rtl" style="text-align:right">في الأسبوع الأول منحنا القياسات أسماء وحسبنا قيمًا جديدة منها. وقبل إضافة القرارات، سنتمهّل ونجعل تلك الفكرة الأساسية صريحة: ما هو المتغير، وماذا يفعل الإسناد، وكيف تتغير القيم بينما تنفّذ <code dir="ltr">Julia</code> سطرًا بعد سطر.</p>
<h3 id="المتغير-يمنح-قيمة-اسما">المتغير يمنح قيمةً اسمًا</h3>
<p dir="rtl" style="text-align:right">المتغير اسم يشير إلى قيمة. والاسم المفيد يتيح لنا الحديث عن معنى العدد بدل تكرار كتابة العدد نفسه:</p>
<p dir="rtl" style="text-align:right">صيغة الإسناد العامة هي <code dir="ltr">name = expression</code>. تحسب <code dir="ltr">Julia</code> التعبير على اليمين وتجعل الاسم على اليسار يشير إلى النتيجة. وأصغر مثال هو <code dir="ltr">x = 2</code>. ويستخدم حساب علمي أغنى أسماءً ذات معنًى متعددة:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [2]:
mass = 2.0        # kg
speed = 3.0       # m/s
momentum = mass * speed   # kg m/s

println(momentum)   # 6.0

6.0


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">6.0</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">علامة المساواة المفردة هي <em>عامل الإسناد</em>. تحسب <code dir="ltr">Julia</code> أولًا التعبير على يمين <code dir="ltr">=</code>؛ ثم تجعل الاسم على اليسار يشير إلى القيمة الناتجة. لذلك يضرب السطر الثالث أولًا <code dir="ltr">mass</code> في <code dir="ltr">speed</code>، ثم يسند النتيجة إلى <code dir="ltr">momentum</code>.</p>
<h4 id="زخم-كتلة-متحركة">زخم كتلة متحركة</h4>
<p dir="rtl" style="text-align:right">يقيس زخم الجسم مقدار الحركة التي يحملها: فالجسم الأثقل والأسرع يحمل زخمًا أكبر. ولنقطة كتلة تتحرك بسرعة أقل بكثير من سرعة الضوء، <span dir="ltr" class="math inline">\(\mathbf{p}=m\mathbf{v}\)</span>، ومقداره <span dir="ltr" class="math inline">\(p=mv\)</span> على امتداد اتجاه الحركة. الكتلة <span dir="ltr" class="math inline">\(m\)</span> بالكيلوغرام، والسرعة <span dir="ltr" class="math inline">\(v\)</span> بالمتر لكل ثانية، والزخم بالكيلوغرام متر لكل ثانية (kg m/s).</p>
<p dir="rtl" style="text-align:right">يمكننا فحص متغير بمجرد كتابة اسمه في خلية دفتر، أو طبعه ضمن رسالة:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">هذه تعبيرات تفاعلية منفصلة. نضع كلًّا منها في خلية مستقلة لتظهر كل النتائج، لا النتيجة الأخيرة فقط.</p>
</div>

In [3]:
momentum

6.0

In [4]:
println("momentum = $(momentum) kg m/s")

momentum = 6.0 kg m/s


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">في الدفتر يعرض السطر الأول <code dir="ltr">6.0</code>؛ أما التقرير المطبوع فهو:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">momentum = 6.0 kg m/s</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<h3 id="تتابع-julia-البرنامج-بالترتيب">تتابع <code dir="ltr">Julia</code> البرنامج بالترتيب</h3>
<p dir="rtl" style="text-align:right">ينفّذ الإسناد حسابًا في تلك اللحظة. وهو لا ينشئ علاقة جبرية دائمة بين المتغيرات:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [5]:
x = 4
y = x + 1       # y is now 5
x = 10          # x changes, but y is still 5

10

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">عندما تنفّذ <code dir="ltr">Julia</code> <code dir="ltr">y = x + 1</code>، تستخدم القيمة الحالية لـ<code dir="ltr">x</code>. وتغيير <code dir="ltr">x</code> لاحقًا لا يعيد حساب <code dir="ltr">y</code> تلقائيًا؛ بل نحتاج إلى تنفيذ <code dir="ltr">y = x + 1</code> من جديد. وهذا الترتيب سطرًا بسطر مهم بشكل خاص في الدفاتر، حيث يمكن تشغيل الخلايا أكثر من مرة أو خارج ترتيبها.</p>
<h3 id="يمكن-إسناد-قيمة-جديدة-لمتغير">يمكن إسناد قيمة جديدة لمتغير</h3>
<p dir="rtl" style="text-align:right">يمكن أن يشير الاسم نفسه إلى قيمة جديدة لاحقًا في البرنامج:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [6]:
temperature = 20.0   # C
temperature = temperature + 0.5

println(temperature)     # 20.5

20.5


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">20.5</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">السطر الثاني ليس معادلة جبرية. تقرأ <code dir="ltr">Julia</code> القيمة القديمة على اليمين، وتضيف <code dir="ltr">0.5</code>، وتسند النتيجة الجديدة إلى الاسم نفسه. وستستخدم الحلقات هذا النمط لتقديم الزمن، أو زيادة عدّاد، أو تحديث موضع محاكى.</p>
<h4 id="يجب-أن-تنقل-الأسماء-المعنى">يجب أن تنقل الأسماء المعنى</h4>
<p dir="rtl" style="text-align:right"><strong>الأسماء</strong>: أسماء المتغيرات حسّاسة لحالة الأحرف، ولا يمكن أن تحتوي فراغات، ويجب استخدامها باتساق. ويمكن استخدام <code dir="ltr">unicode</code>، ولا سيّما الرموز اليونانية (مثل كتابة <code dir="ltr">\phi&lt;Tab&gt;</code> للحصول على <span dir="ltr" class="math inline">\(\phi\)</span>). نفضّل الأسماء ذات المعنى على أسماء غامضة مثل <code dir="ltr">a</code> و<code dir="ltr">b</code> و<code dir="ltr">c</code>. <strong>الوحدات</strong>: تخزّن <code dir="ltr">Julia</code> عادة أعدادًا عادية دون وحداتها. لكن في الفيزياء يُستحسن تتبّعها. ويمكن إضافة لواحق وحدية إلى الأسماء، مثل <code dir="ltr">m_kg</code>، أو بدلًا من ذلك، كما نفعل هنا، تسجيل الوحدة في تعليق قصير في نهاية السطر حيث تُعرَّف القيمة، مثل <code dir="ltr">time = 0.35 # s</code> و<code dir="ltr">height = 1.20 # m</code> و<code dir="ltr">g = 9.81 # m/s^2</code>؛ وهذا ليس إلزاميًا لكنه مفيد جدًا لتجنّب أخطاء الوحدات. وفي واجب المكافأة 2 (أي أنه ليس جزءًا من مادة المقرر الرسمية) سنرى أننا نستطيع استخدام حزمة <code dir="ltr">Unitful</code> التي تضيف وحدات إلى المتغيرات وتمكّننا من تتبّعها. ويجب أيضًا ذكر الوحدات في توثيق الدوال أو توثيق الواجهة، وتوضيحها عند مواقع الاستدعاء عندما تكون ملتبسة.</p>
<h3 id="للقيم-أنواع">للقيم أنواع</h3>
<p dir="rtl" style="text-align:right">يمكن للمتغيرات أن تسمّي أنواعًا مختلفة من القيم. وقد استخدمنا في الأسبوع الأول عددًا صحيحًا للعدّ، وقياسات عشرية، وتسميات نصية. وهذا الأسبوع نضيف مقارنات منطقية ونجعل أربعة تمييزات صريحة، لأن برنامجًا يقرأ مدخلات ويختار فرعًا ويقرر متى يتوقف يجب أن يعرف نوع القيمة التي يتعامل معها.</p>
<p dir="rtl" style="text-align:right"><em>النوع</em> فئة من القيم تحدّد كيف تخزّنها <code dir="ltr">Julia</code> وأي العمليات تكون ذات معنى. فمثلًا، لـ<code dir="ltr">5</code> و<code dir="ltr">0.35</code> و<code dir="ltr">"height"</code> و<code dir="ltr">true</code> صيغ مختلفة، وهي تمثّل عددًا صحيحًا وعددًا عشريًا ونصًا وقيمة صواب/خطأ.</p>
<h3 id="أربعة-أنواع-قيم-بسيطة">أربعة أنواع قيم بسيطة</h3>
<h4 id="الأنواع-الأربعة-التي-نحتاجها-الآن">الأنواع الأربعة التي نحتاجها الآن</h4>
<table>
<thead>
<tr>
<th style="text-align: left;">نوع <code dir="ltr">Julia</code></th>
<th style="text-align: left;">مثال</th>
<th style="text-align: left;">الدور العلمي المعتاد</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;"><code dir="ltr">Int</code></td>
<td style="text-align: left;"><code dir="ltr">n = 5</code></td>
<td style="text-align: left;">العدّ وفهارس الحلقات</td>
</tr>
<tr>
<td style="text-align: left;"><code dir="ltr">Float64</code></td>
<td style="text-align: left;"><code dir="ltr">t = 0.35 # s</code></td>
<td style="text-align: left;">القياسات والكميات المحسوبة</td>
</tr>
<tr>
<td style="text-align: left;"><code dir="ltr">String</code></td>
<td style="text-align: left;"><code dir="ltr">label = "height [m]"</code></td>
<td style="text-align: left;">نص المدخلات والتسميات والرسائل</td>
</tr>
<tr>
<td style="text-align: left;"><code dir="ltr">Bool</code></td>
<td style="text-align: left;"><code dir="ltr">ready = true</code></td>
<td style="text-align: left;">أجوبة الصواب/الخطأ عن الشروط</td>
</tr>
</tbody>
</table>
<p dir="rtl" style="text-align:right">تتيح لنا الدالة <code dir="ltr">typeof(value)</code> فحص نوع قيمة. ونستخدمها للتشخيص، لا بديلًا عن فهم معنى القيمة.</p>
<p dir="rtl" style="text-align:right">في حاسبات المقرر العادية ذات 64 بت، تطبع <code dir="ltr">typeof(5)</code> النص <code dir="ltr">Int64</code>. واسم <code dir="ltr">Int</code> في <code dir="ltr">Julia</code> يعني حجم العدد الصحيح المعتاد للحاسوب، وهو اسم بديل لـ<code dir="ltr">Int64</code> في هذه الحاسبات.</p>
<p dir="rtl" style="text-align:right">وترتبط هذه الأنواع مباشرةً بالأسبوع الأول:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [7]:
current = [0.10, 0.20, 0.30, 0.40, 0.50]   # A
n = length(current)                 # Int
I_bar = sum(current) / n            # Float64, A
axis_label = "current I [A]"    # String
analysis_ready = true           # Bool

println(typeof(n))
println(typeof(I_bar))
println(typeof(axis_label))
println(typeof(analysis_ready))

Int64


Float64
String
Bool


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">يمكن لعدد صحيح وعدد عائم أن يشاركا في حساب واحد. وتحوّل <code dir="ltr">Julia</code> العدد الصحيح عند الحاجة:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">هذه تعبيرات تفاعلية منفصلة. نضع كلًّا منها في خلية مستقلة لتظهر كل النتائج، لا النتيجة الأخيرة فقط.</p>
</div>

In [8]:
2 * 0.35       # 0.7, a Float64

0.7

In [9]:
5 / 2          # 2.5, a Float64

2.5

In [10]:
Float64(5)     # 5.0

5.0

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">ولهذه المرحلة تكفي هذه القواعد العملية. لا نحتاج إلى تعليقات الأنواع، ولا إلى الأنواع المجردة، ولا إلى تفاصيل تخزين الأعداد العائمة.</p>
<h4 id="النوع-ليس-وحدة-فيزيائية">النوع ليس وحدة فيزيائية</h4>
<p dir="rtl" style="text-align:right">في <code dir="ltr">t = 0.35 # s</code>، النوع في <code dir="ltr">Julia</code> هو <code dir="ltr">Float64</code>؛ أما الوحدة الفيزيائية فهي الثانية. والوحدة تعيش في التعليق وفي توثيق الواجهة المعنيّ، لذا يمكن لـ<code dir="ltr">Float64</code> آخر أن يمثّل المتر أو الكيلوغرام. ففحص النوع وتحليل الأبعاد يجيبان عن سؤالين مختلفين، ونحتاج إليهما معًا.</p>
<h3 id="تنتج-المقارنات-قيما-منطقية">تنتج المقارنات قيمًا منطقية</h3>
<p dir="rtl" style="text-align:right">تحتاج البرامج إلى أسئلة بنعم/لا للتحقق من البيانات واختيار ما تفعله بعد ذلك. وللمقارنة الصيغة <code dir="ltr">left_value operator right_value</code> وتنتج قيمة <code dir="ltr">Bool</code>: إما <code dir="ltr">true</code> أو <code dir="ltr">false</code>. نبدأ بأسئلة مباشرة أجوبتها واضحة:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">هذه تعبيرات تفاعلية منفصلة. نضع كلًّا منها في خلية مستقلة لتظهر كل النتائج، لا النتيجة الأخيرة فقط.</p>
</div>

In [11]:
2 < 3          # true

true

In [12]:
2 == 3         # false

false

In [13]:
5 != 4         # true

true

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">عوامل المقارنة الرئيسية هي <code dir="ltr">&lt;</code> و<code dir="ltr">&lt;=</code> و<code dir="ltr">&gt;</code> و<code dir="ltr">&gt;=</code> و<code dir="ltr">==</code> (يساوي) و<code dir="ltr">!=</code> (لا يساوي). وعلامة المساواة المفردة <code dir="ltr">=</code> تسند قيمة؛ أما المزدوجة <code dir="ltr">==</code> فتطرح سؤالًا.</p>
<p dir="rtl" style="text-align:right">يمكننا دمج القيم المنطقية عندما لا تكفي مقارنة واحدة. والعوامل <code dir="ltr">&amp;&amp;</code> و<code dir="ltr">||</code> و<code dir="ltr">!</code> تعني و، أو، وليس:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">هذه تعبيرات تفاعلية منفصلة. نضع كلًّا منها في خلية مستقلة لتظهر كل النتائج، لا النتيجة الأخيرة فقط.</p>
</div>

In [14]:
true && false      # false: both must be true

false

In [15]:
true || false      # true: at least one is true

true

In [16]:
!true              # false

false

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">ويمكن للصيغ نفسها أن تعبّر عن قاعدة صلاحية فيزيائية أغنى:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [17]:
t = 0.35   # s
z0 = 1.20   # m
nonnegative_time = t >= 0
valid_height = z0 >= 0
valid_input = nonnegative_time && valid_height

true

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<h3 id="قرار-أول-باستخدام-if">قرار أول باستخدام <code dir="ltr">if</code></h3>
<p dir="rtl" style="text-align:right">لا يكون حساب جواب منطقي مفيدًا إلا إذا استطاع البرنامج أن يتصرف بناءً عليه. وعبارة <code dir="ltr">if</code> شرطية: تنفّذ كتلة واحدة عندما يكون شرطها <code dir="ltr">true</code>. وكتلة <code dir="ltr">else</code> الاختيارية تعمل عندما يكون الشرط <code dir="ltr">false</code>. والصيغة العامة هي:</p>
</div>

<div dir="rtl" style="text-align:right">
<h3 id="قالب-الصياغة">قالب الصياغة</h3>
<p dir="rtl" style="text-align:right">هذا هو الشكل العام؛ و<code dir="ltr">condition</code> عنصر نائب. نقرؤه هنا ثم ننفّذ المثال الملموس الذي يليه.</p>
</div>

In [ ]:
if condition
    # commands when the condition is true
else
    # commands when the condition is false
end

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">كل شرطية تنتهي بـ<code dir="ltr">end</code>. وإذا لم يكن ينبغي أن يحدث شيء في الحالة الخطأ، نحذف كتلة <code dir="ltr">else</code>. وهنا أصغر مثال ثنائي الاتجاه:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [18]:
temperature = 25.0   # C
if temperature > 30
    println("hot")
else
    println("not hot")
end

not hot


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">not hot</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">قد يجعل المتابعة حسابًا بلا معنًى في بعض الحالات. وتوفّر <code dir="ltr">Julia</code> توقفًا مقصودًا بهذه الصيغة:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

<div dir="rtl" style="text-align:right">
<h3 id="خطأ-مقصود-يلاحظ-بأمان">خطأ مقصود يُلاحَظ بأمان</h3>
<p dir="rtl" style="text-align:right">يتوقف الفصل هنا عمدًا عند حساب. غلاف <code dir="ltr">try</code>/<code dir="ltr">catch</code> المرفق يطبع الخطأ المتوقّع ويسمح لـ Run All بالمتابعة؛ أما الأخطاء غير المتوقعة فتُعاد رفعها. نركّز على التعليمة الفاشلة، لا على تعلّم الغلاف في هذه المرحلة.</p>
</div>

In [19]:
try
    throw(ArgumentError("message"))
catch err
    err isa ArgumentError || rethrow()
    showerror(stdout, err)
    println()
end

ArgumentError: message


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">إنه يوقف التنفيذ ويبلّغ عن السبب. وسنستخدمه بعد تحويل فاشل وبعد مدخلات غير صالحة فيزيائيًا.</p>
<h2 id="الحركية-متعددة-القطع-كقرار">الحركية متعددة القطع كقرار</h2>
<h4 id="التسارع-الموجب-مقابل-مقدار-الجاذبية-الموجب">التسارع الموجَّب مقابل مقدار الجاذبية الموجب</h4>
<p dir="rtl" style="text-align:right">الارتفاع موجب نحو الأعلى، لذا يكون تسارع الجاذبية <span dir="ltr" class="math display">\[g=-9.81\ \mathrm{m/s^2}.\]</span> ولارتفاع ابتدائي <span dir="ltr" class="math inline">\(z_0\)</span> وسرعة <span dir="ltr" class="math inline">\(v_0\)</span>، <span dir="ltr" class="math display">\[z_{\mathrm{free}}(t)=z_0+v_0t+\frac{1}{2}gt^2.\]</span> وللحدود الثلاثة وحدات المتر. وهذا الفحص البُعدي مستقل عن كون قيمها في الكود من النوع <code dir="ltr">Float64</code>.</p>
<p dir="rtl" style="text-align:right">نموذجنا دون ارتداد لا يسمح للكرة بمواصلة الحركة تحت الأرض: <span dir="ltr" class="math display">\[z(t)=\begin{cases}
z_{\mathrm{free}}(t),&amp;z_{\mathrm{free}}(t)&gt;0,\\
0,&amp;z_{\mathrm{free}}(t)\leq0.
\end{cases}\]</span></p>
<p dir="rtl" style="text-align:right">ويتحوّل التعريف متعدد القطع مباشرةً إلى شرطية:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [20]:
t = 0.35   # s
z0 = 1.20   # m
v0 = 0.0    # m/s
g = -9.81   # m/s^2

if t < 0 || z0 < 0
    throw(ArgumentError("time and initial height must be non-negative"))
end

z_free = z0 + v0 * t + 0.5 * g * t^2   # m

if z_free > 0
    z = z_free
    state = "in air"
else
    z = 0.0
    state = "on the ground"
end

println("at t = $(t) s: z = $(round(z, digits=3)) m ($(state))")

at t = 0.35 s: z = 0.599 m (in air)


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">at t = 0.35 s: z = 0.599 m (in air)</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">عند <span dir="ltr" class="math inline">\(t=0.35\ \mathrm{s}\)</span> يعرض هذا البرنامج نحو <span dir="ltr" class="math inline">\(z=0.599\ \mathrm{m}\)</span>. وتجتاز النتيجة ثلاثة فحوص: الزمن داخل مجال النموذج، والارتفاع أدنى من قيمته الابتدائية، ولكل حد في معادلة الموضع وحدات المتر.</p>
<h3 id="عدة-فروع">عدة فروع</h3>
<p dir="rtl" style="text-align:right">نستخدم <code dir="ltr">elseif</code> عندما تكون أكثر من حالتين متنافيتين وينبغي اختبارهما بالترتيب. والصيغة العامة هي <code dir="ltr">if</code>، يتبعها أي عدد من فروع <code dir="ltr">elseif</code>، ثم <code dir="ltr">else</code> اختياري، ثم <code dir="ltr">end</code> واحد. ولا يعمل إلا أول فرع صحيح. ويعرض مثال صغير لتصنيف الإشارة هذه البنية:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [21]:
value = -2
if value < 0
    description = "negative"
elseif value == 0
    description = "zero"
else
    description = "positive"
end

"negative"

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">وتعطي البنية نفسها تصنيفًا أكثر فائدة للارتفاع:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [22]:
if z == 0
    state = "on the ground"
elseif z < 0.20
    state = "close to the ground"
else
    state = "in air"
end

"in air"

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">تكون القرارات المتداخلة ضرورية أحيانًا، لكن سلسلة <code dir="ltr">if</code>/<code dir="ltr">elseif</code>/<code dir="ltr">else</code> المسطّحة أسهل في الفحص عندما تعبّر عن المنطق نفسه.</p>
<h4 id="ثلاث-فئات-من-الأخطاء">ثلاث فئات من الأخطاء</h4>
<ul>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right"><strong>خطأ صياغي:</strong> لا تستطيع <code dir="ltr">Julia</code> تحليل البرنامج، مثلًا بسبب <code dir="ltr">end</code> مفقود أو قوس إغلاق ناقص.</p></li>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right"><strong>خطأ زمن التشغيل:</strong> يبدأ البرنامج ثم يصل إلى عملية غير صالحة، مثلًا رفع <code dir="ltr">String</code> إلى قوة.</p></li>
<li dir="rtl" style="text-align:right"><p dir="rtl" style="text-align:right"><strong>خطأ منطقي:</strong> يعمل البرنامج لكنه ينفّذ الخوارزمية أو الفيزياء الخطأ، مثلًا استخدام <span dir="ltr" class="math inline">\(g\)</span> موجب بينما الارتفاع موجب نحو الأعلى، أو عدم تقديم الزمن في حلقة <code dir="ltr">while</code>.</p></li>
</ul>
<p dir="rtl" style="text-align:right">نقرأ الخطأ الدقيق ورقم السطر، ونعزل أصغر كتلة فاشلة، ونفحص القيم والأنواع، ونتحقق من الأبعاد والإشارات، ونختبر حالات أجوبتها معروفة سلفًا.</p>
<p dir="rtl" style="text-align:right">نجعل البرامج الآن تختار بين حالات فيزيائية مختلفة. وقبل الوثوق بأي حساب، علينا أن نعرف نوع القيمة التي بين أيدينا، وأن نحوّل عن قصد، وأن نذكر المجال الذي ينطبق فيه النموذج.</p>
<h2 id="للقيم-أنواع">للقيم أنواع</h2>
<p dir="rtl" style="text-align:right">تحمل قيم <code dir="ltr">Julia</code> أنواعًا. فالأعداد الصحيحة والأعداد العشرية والنصوص وقيم الصواب تخدم أغراضًا مختلفة.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [23]:
println(3, "  ->  ", typeof(3))
println(3.0, "  ->  ", typeof(3.0))
println("3", "  ->  ", typeof("3"))
println(3 < 4, "  ->  ", typeof(3 < 4))

3  ->  Int64
3.0  ->  Float64
3  ->  String
true  ->  

Bool


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">3  -&gt;  Int64
3.0  -&gt;  Float64
3  -&gt;  String
true  -&gt;  Bool</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">قد يعتمد نوع العدد الصحيح الدقيق على الحاسوب؛ لكن التمييز المهم هو بين <code dir="ltr">Int</code> والعائم و<code dir="ltr">String</code> و<code dir="ltr">Bool</code>.</p>
<p dir="rtl" style="text-align:right">يمكن لعدد صحيح وعدد عائم أن يختلطا في الحساب العادي؛ وترفع <code dir="ltr">Julia</code> النوع عند الحاجة. أما النص فليس عددًا ضمنًا. ونختار التحويلات صراحةً في الاتجاهين.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [24]:
println(2 + 0.5)
text = "12.5"
mass = parse(Float64, text)   # kg
println(mass + 0.5)
println(string(mass, " kg"))
println(Int(7.0))

2.5
13.0
12.5 kg
7


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">2.5
13.0
12.5 kg
7</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">يطلق <code dir="ltr">parse(Float64, "twelve")</code> خطأً لأن النص ليس عددًا عشريًا. ويطلق <code dir="ltr">Int(7.2)</code> خطأً أيضًا: فالتحويل إلى <code dir="ltr">Int</code> مسموح فقط عندما تكون القيمة صحيحة تمامًا. وهذا يحمينا من التخلص من المعلومة بصمت.</p>
<h2 id="القيم-المنطقية-والقرارات-التدريجية">القيم المنطقية والقرارات التدريجية</h2>
<p dir="rtl" style="text-align:right">تنتج المقارنات قيمًا منطقية. نسأل أولًا سؤالًا واحدًا، ثم نختار بين حالتين، ثم نضيف حالة ثالثة بـ<code dir="ltr">elseif</code>.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [25]:
temperature = -2.0   # C
println(temperature < 0)
if temperature < 0
    println("below freezing")
else
    println("not below freezing")
end

true
below freezing


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">true
below freezing</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [26]:
if temperature < 0
    state = "ice"
elseif temperature == 0
    state = "melting-point"
else
    state = "liquid"
end
println(state)

ice


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">ice</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><code dir="ltr">ice</code>. ولا يعمل من سلسلة <code dir="ltr">if</code>/<code dir="ltr">elseif</code>/<code dir="ltr">else</code> إلا فرع واحد بالضبط. و<code dir="ltr">=</code> يسند؛ و<code dir="ltr">==</code> يقارن.</p>
<p dir="rtl" style="text-align:right">نجمع الشروط بـ<code dir="ltr">&amp;&amp;</code> (كلاهما) و<code dir="ltr">||</code> (أحدهما على الأقل) و<code dir="ltr">!</code> (ليس). وغالبًا ما يسأل فحص الصلاحية هل القيمة عدد منتهٍ. ويعيد <code dir="ltr">isfinite(value)</code> القيمة <code dir="ltr">true</code> للأعداد المنتهية العادية و<code dir="ltr">false</code> لـ<code dir="ltr">NaN</code> و<code dir="ltr">Inf</code> و<code dir="ltr">-Inf</code>، وهي ليست قياسات قابلة للاستخدام. ويمكننا دمج ذلك الفحص مع شرط فيزيائي:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [27]:
is_finite_positive = isfinite(2.0) && 2.0 > 0
println(is_finite_positive)
println(!is_finite_positive || 2.0 == 2)

true
true


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">true
true</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<h2 id="الزوايا-والمنطق-والتحقق-قبل-الحساب">الزوايا والمنطق والتحقق قبل الحساب</h2>
<h4 id="الشرط-سؤال-رياضي-دقيق">الشرط سؤال رياضي دقيق</h4>
<p dir="rtl" style="text-align:right">يكون <code dir="ltr">a &amp;&amp; b</code> صحيحًا فقط عندما يكون الشرطان صحيحين؛ ويكون <code dir="ltr">a || b</code> صحيحًا عندما يكون أحدهما على الأقل صحيحًا. ويكون <code dir="ltr">xor(a,b)</code> صحيحًا عندما يكون أحدهما بالضبط صحيحًا. وتفيد هذه القواعد في اختيار حالات متنافية، لكن على النموذج أن يحدّد ما يحدث عند المساواة. ومعنى متباينة متسلسلة مثل <span dir="ltr" class="math inline">\(0\le\theta&lt;90^\circ\)</span> أن يتحقق الحدّان معًا؛ ويمكننا كتابتها بمقارنتين موصولتين بـ<code dir="ltr">&amp;&amp;</code>.</p>
<p dir="rtl" style="text-align:right">تحتاج الزاوية إلى وحدة صريحة. يتوقّع <code dir="ltr">sin</code> و<code dir="ltr">cos</code> الراديان، بينما يتوقّع <code dir="ltr">sind</code> و<code dir="ltr">cosd</code> الدرجات؛ ويحوّل <code dir="ltr">rad2deg</code> زاوية من الراديان إلى الدرجات. أما الصيغة الثنائية <code dir="ltr">atan(y, x)</code> فتعيد زاوية النقطة <span dir="ltr" class="math inline">\((x,y)\)</span> واعيةً بالربع: فهي تستخدم الإشارتين معًا، لا نسبتهما فقط، فتستطيع التمييز بين اتجاهين متعاكسين. وفيما يلي مثال صغير لكلٍّ منها:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [28]:
println(true && false)
println(true || false)
println(xor(true, true))
println(xor(true, false))
println(sin(pi / 6))
println(sind(30.0))
println(round(rad2deg(atan(1.0, -1.0)); digits=1))

false
true
false
true
0.49999999999999994
0.5
135.0


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">false
true
false
true
0.49999999999999994
0.5
135.0</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">تقع النقطة <span dir="ltr" class="math inline">\((-1,1)\)</span> عند <span dir="ltr" class="math inline">\(135^\circ\)</span>، لا عند <span dir="ltr" class="math inline">\(-45^\circ\)</span>؛ ولا تستطيع الصيغة ذات الأرجومنت الواحد <code dir="ltr">atan(y/x)</code> التمييز بينهما. وعند نقطة الأصل لا يوجد اتجاه فيزيائي؛ وعلى أي تطبيق أن يذكر اصطلاحًا. والفروق الصغيرة بين حسابات فاصلة مكافئة ليست فيزياءً مختلفة.</p>
<h4 id="تحويل-النص-والتحقق-الفيزيائي-عمليتان-منفصلتان">تحويل النص والتحقق الفيزيائي عمليتان منفصلتان</h4>
<p dir="rtl" style="text-align:right">يعيد <code dir="ltr">readline()</code> نصًا. ويحوّل <code dir="ltr">parse(Float64, text)</code> النص العددي الصالح إلى عدد؛ والنص المشوّه يطلق خطأً. ولا يكفي نجاح التحويل: فـ<code dir="ltr">NaN</code> و<code dir="ltr">Inf</code> ليست قياسات منتهية، والكتلة السالبة المنتهية خارج نموذجنا. نتحقق قبل جذر أو قسمة أو بناء مثلثي. ويوقف <code dir="ltr">throw(ArgumentError("message"))</code> الحساب؛ أما مجرد طباعة رسالة خطأ فلا توقفه.</p>
<p dir="rtl" style="text-align:right">قبل اختبار مجال فيزيائي، نتأكد أن القيمة منتهية: يكون <code dir="ltr">isfinite(value)</code> صحيحًا فقط للأعداد المنتهية العادية، وخطأً لـ<code dir="ltr">NaN</code> أو <code dir="ltr">Inf</code>. وتحوّل الشيفرة العددية التالية نصًا وتتحقق من نتيجتها ككتلة:</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [29]:
text = "2.5"  # reproducible example of text supplied by readline()
mass = parse(Float64, text)   # kg
if !isfinite(mass) || mass <= 0
    throw(ArgumentError("mass must be finite and positive"))
end
println("mass = ", mass, " kg")

mass = 2.5 kg


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">mass = 2.5 kg</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">وتبقى أمثلة <code dir="ltr">tryparse</code>/<code dir="ltr">nothing</code> السابقة بأسلوب التعافي للقراءة الاختيارية في معالجة المدخلات. أما المسار الأساسي هذا الأسبوع فيستخدم <code dir="ltr">parse</code> وفحوص المجال الفيزيائي الصريحة؛ ولا يتطلب أي منهما تعريف دالة.</p>
<h2 id="نموذج-احتكاك-متعدد-القطع-على-منحدر">نموذج احتكاك متعدد القطع على منحدر</h2>
<h4 id="القوى-والإشارات-والافتراضات">القوى والإشارات والافتراضات</h4>
<p dir="rtl" style="text-align:right">نأخذ <span dir="ltr" class="math inline">\(+x\)</span> نحو الأعلى بمحاذاة منحدر ثابت و<span dir="ltr" class="math inline">\(+y\)</span> عموديًا عليه. ولجسم كتلته <span dir="ltr" class="math inline">\(m\)</span> (كغ) وزاوية <span dir="ltr" class="math inline">\(\theta\)</span> (بالراديان في <code dir="ltr">sin</code>/<code dir="ltr">cos</code>) ومقدار جاذبية موجب <span dir="ltr" class="math inline">\(g\)</span> (m/s<span dir="ltr" class="math inline">\(^2\)</span>)، للجاذبية مركّبة نزولية <span dir="ltr" class="math inline">\(mg\sin\theta\)</span> وقوة عمودية <span dir="ltr" class="math inline">\(N=mg\cos\theta\)</span>. ويمكن للاحتكاك السكوني أن يعارض الحركة الوشيكة نحو الأسفل حتى <span dir="ltr" class="math inline">\(f_{s,\max}=\mu_sN\)</span>. وإذا لم يستطع التثبيت، يعمل الاحتكاك الحركي <span dir="ltr" class="math inline">\(f_k=\mu_kN\)</span> نحو الأعلى ويكون <span dir="ltr" class="math inline">\(a=(-mg\sin\theta+\mu_kN)/m\)</span>. نفترض احتكاك كولوم وتلامسًا صلبًا ولا مقاومة هواء و<span dir="ltr" class="math inline">\(0\le\theta&lt;90^\circ\)</span>. ولا يصف النموذج التدحرج أو تغيّر التلامس أو احتكاكًا يعتمد على السرعة. وعند <span dir="ltr" class="math inline">\(\theta=0\)</span> يكون الاحتكاك السكوني المطلوب صفرًا؛ وإذا كان <span dir="ltr" class="math inline">\(\mu_k=0\)</span>، يعطي الانزلاق <span dir="ltr" class="math inline">\(a=-g\sin\theta\)</span>.</p>
<p dir="rtl" style="text-align:right">نتحقق قبل استخدام أي مدخل. يوقف <code dir="ltr">throw(ArgumentError(...))</code> الحساب فورًا ويخبر مستدعيه أن شرطًا موثّقًا في المدخل قد فشل. وهو يختلف عن <code dir="ltr">println</code> الذي يكتفي بعرض نص ثم يكمل. ويسأل اختبار العضوية <code dir="ltr">value isa T</code> هل تنتمي القيمة إلى النوع <code dir="ltr">T</code>؛ ويسمّي <code dir="ltr">Real</code> عائلة <code dir="ltr">Julia</code> للأنواع الحقيقية، التي تضم الأعداد الصحيحة والعائمة العادية. ويحتاج <code dir="ltr">Bool</code> إلى عناية خاصة هنا: فبما أن <code dir="ltr">true isa Real</code> صحيحة أيضًا، فإن فحص <code dir="ltr">Real</code> وحده سيقبل <code dir="ltr">true</code> ككتلة، لذا نرفضه صراحةً.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>قبل التنفيذ:</strong> نتوقّع القيمة أو النص أو الرسم الموصوف أعلاه. هذا هو المثال المقابل من الفصل المكتوب.</p>
</div>

In [30]:
mass = 2.0      # kg
angle = 30.0    # deg
 μ_s = 0.60
 μ_k = 0.40
g = 9.81        # m/s^2
if !(mass isa Real) || mass isa Bool || !isfinite(mass) || mass <= 0
    throw(ArgumentError("mass must be finite, positive, and non-Boolean"))
elseif !isfinite(angle) || !(0 <= angle < 90) || μ_k < 0 || μ_k > μ_s
    throw(ArgumentError("use 0 <= angle < 90 and 0 <= μ_k <= μ_s"))
end
 θ = deg2rad(angle)   # rad
normal = mass * g * cos(θ)   # N
needed_static = mass * g * sin(θ)   # N
if needed_static <= μ_s * normal
    state = "stuck"
    acceleration = 0.0   # m/s^2
else
    state = "sliding"
    acceleration = (-needed_static + μ_k * normal) / mass   # m/s^2
end
println(state, ", a = ", acceleration, " m/s^2")

stuck, a = 0.0 m/s^2


<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><strong>تحقّق:</strong> نقارن النتيجة بالشرح الوارد أدناه. تعرض الخلية تلقائيًا تعبيرها الأخير فقط؛ أما <code dir="ltr">println</code> فتكتب النص عمدًا. نعيد تشغيل الخلايا السابقة عند تغيير أي مدخل.</p>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right">مخرجات مرجعية في الملاحظات المكتوبة (للمقارنة مع مخرجات الخلية الفعلية):</p>
<pre dir="ltr" style="text-align:left" class="text"><code dir="ltr">stuck, a = 0.0 m/s^2</code></pre>
</div>

<div dir="rtl" style="text-align:right">
<p dir="rtl" style="text-align:right"><code dir="ltr">stuck, a = 0.0 m/s^2</code>. هنا <span dir="ltr" class="math inline">\(mg\sin30^\circ=9.81\)</span> N و<span dir="ltr" class="math inline">\(\mu_sN\approx10.19\)</span> N، فالاحتكاك السكوني كافٍ. وإذا غيّرنا <code dir="ltr">μ_s</code> وحده إلى <code dir="ltr">0.40</code>، تكون النتيجة <code dir="ltr">sliding</code> مع تسارع سالب: أي نحو الأسفل في اصطلاحنا المعلن. والمساواة تنتمي إلى فرع التثبيت، وهو اصطلاح العتبة الصريح عندنا.</p>
<h2 id="التمرين-قوة-سحب-قد-تعكس-حركة-وشيكة">التمرين: قوة سحب قد تعكس حركة وشيكة</h2>
<p dir="rtl" style="text-align:right">نضيف إلى نموذج المنحدر قوة سحب معروفة نحو الأعلى <span dir="ltr" class="math inline">\(P\)</span>. وقبل اختيار إشارة الاحتكاك، نحسب النزعة بلا احتكاك <span dir="ltr" class="math inline">\(T=P-mg\sin\theta\)</span>. فالنزعة الموجبة <span dir="ltr" class="math inline">\(T\)</span> تعني حركة وشيكة نحو الأعلى، والسلبية تعني حركة وشيكة نحو الأسفل. ويعارض الاحتكاك تلك النزعة، ويمكن للاحتكاك السكوني أن يثبّت متى كان <span dir="ltr" class="math inline">\(|T|\le\mu_sN\)</span>.</p>
<p dir="rtl" style="text-align:right">نصل قادرين على قراءة سلسلة <code dir="ltr">if</code>/<code dir="ltr">elseif</code>/<code dir="ltr">else</code> وتحويل الدرجات إلى راديان. وفي الجزء الموجَّه القصير نتوقّع الاتجاه، ونتتبّع فرعًا واحدًا، ونختبر <span dir="ltr" class="math inline">\(T=0\)</span>. ثم نبدأ واجب خلوص الجدار المستقل. والفكرة الجوهرية طريقة قابلة للتكرار: نتحقق من المدخلات، ونصوغ شرطًا فيزيائيًا، ونحسب الحالة المنطبقة، ونفحص حدّها.</p>
<h2 id="الواجب-الأساسي-هل-يعبر-المقذوف-جدارا">الواجب الأساسي: هل يعبر المقذوف جدارًا؟</h2>
<p dir="rtl" style="text-align:right">نطلق من مستوى الأرض بسرعة <span dir="ltr" class="math inline">\(v_0\)</span> وزاوية <span dir="ltr" class="math inline">\(\theta\)</span> ومقدار جاذبية <span dir="ltr" class="math inline">\(g&gt;0\)</span>، مع <span dir="ltr" class="math inline">\(+x\)</span> نحو الجدار و<span dir="ltr" class="math inline">\(+y\)</span> نحو الأعلى. وفي نموذج الجاذبية المنتظمة دون مقاومة هواء، <span dir="ltr" class="math display">\[v_x=v_0\cos\theta,\quad t_w=x_w/v_x,\quad y_w=v_0\sin\theta\,t_w-\tfrac12gt_w^2,\quad v_{y,w}=v_0\sin\theta-gt_w.\]</span> ونشترط <span dir="ltr" class="math inline">\(v_x&gt;0\)</span> قبل القسمة. نقارن <span dir="ltr" class="math inline">\(y_w\)</span> بارتفاع الجدار، ونستخدم <span dir="ltr" class="math inline">\(v_{y,w}\)</span> لتصنيف الصعود أو الهبوط. والمساواة في الارتفاع تُحتسب عبورًا؛ والسرعة الرأسية الصفرية تنتمي إلى فئة الهبوط المنشورة. وقد يكون الارتفاع المحسوب في الطيران الحر سالبًا: وبالنسبة لارتفاع جدار غير سالب يعني ذلك فشل العبور، لا استمرار الحركة فيزيائيًا عبر الأرض.</p>
<p dir="rtl" style="text-align:right">نكتب أولًا تنبؤًا لحالة عادية وأخرى حدّية. ثم نطبّق الشيفرة العددية، ونتحقق من مجالها الموثّق، ونقارن التسميات والقيمة المنطقية بدقة. ونتحقق من الارتفاعات تقريبيًا بالتحمّل المنشور. وأخيرًا نشرح كيف يختلف قرار العبور عن قرار الحالة الرأسية: فقد يعبر المقذوف وهو هابط. وورقة الواجب هي المكان الذي تُذكر فيه أسماء المدخلات الدقيقة والاختبارات واصطلاح غلاف الإرجاع وتعليمات التسليم؛ ولا حاجة إلى دوال أو حلقات يعرّفها الطالب.</p>
<h2 id="مكافأة-اختيارية-الوحدات-وأطوار-التيار-المتردد">مكافأة اختيارية: الوحدات وأطوار التيار المتردد</h2>
<p dir="rtl" style="text-align:right">هذه المكافأة اختيارية، وليست شرطًا مسبقًا لعمل لاحق، وغير ممتحنة. وتساوي <strong>ثلث نقطة من العلامة النهائية</strong>.</p>
<p dir="rtl" style="text-align:right">تطوّر القراءة المستقلة الأعداد المركبة والمقدار والطور والمرافق وعلاقة أويلر قبل استخدام <code dir="ltr">cis</code>. نختار سعات الذروة واصطلاح الزمن <span dir="ltr" class="math inline">\(\exp(+i\omega t)\)</span>، ونضع طور الجهد صفرًا، ونبني <span dir="ltr" class="math inline">\(Z=|Z|\exp(i\phi_Z)\)</span>، ونحسب <span dir="ltr" class="math inline">\(I=V/Z\)</span>. والتيار القابل للقياس هو الجزء الحقيقي من <span dir="ltr" class="math inline">\(I\exp(i\omega t)\)</span>؛ فالعدد المركب تمثيل، لا تيار مقيس إضافي.</p>
<p dir="rtl" style="text-align:right">نتبع أمثلة <code dir="ltr">Unitful</code> المُعطاة، ونتحقق من أن قسمة الجهد على المقاومة لها وحدة التيار، ونفحص مقاومًا بطور صفري قبل تجربة طور ربع دورة. ثم نقدّم الشيفرة العددية التي تسند التيار المركب بالأمبير ونشغّل فحوص الوحدات المكافئة والمدخلات غير الصالحة المنشورة. ونشرح لماذا يعطي طور معاوقة موجب طور تيار سالبًا. هذا نموذج أحادي التردد خطي في الحالة المستقرة؛ والانتقالات العابرة والمكوّنات غير الخطية خارج نطاقه.</p>
<h2 id="استكشاف-إضافي-الانعكاس-الداخلي-الكلي">استكشاف إضافي: الانعكاس الداخلي الكلي</h2>
<p dir="rtl" style="text-align:right">نستخدم وصفة قانون سنيل المُعطاة <span dir="ltr" class="math inline">\(n_1\sin\theta_1=n_2\sin\theta_2\)</span>، مع قرائن انكسار موجبة وزوايا مقيسة من العمودي على السطح الفاصل. وعندما يكون <span dir="ltr" class="math inline">\(n_1&gt;n_2\)</span>، يعطي السقوط بعد <span dir="ltr" class="math inline">\(\theta_c=\arcsin(n_2/n_1)\)</span> انعكاسًا داخليًا كليًا. ونصنّف الأمثلة دون الحرجة والحرجة وفوق الحرجة قبل حساب جيب معكوس؛ فحساب <code dir="ltr">asin</code> لعدد أكبر من واحد دون تفكير ليس حسابًا فيزيائيًا.</p>
<p dir="rtl" style="text-align:right">نرسم انعكاسات مثالية متكررة في ليف مستقيم ونغيّر زاوية الإطلاق. ونسجّل الخيارات التي تبقى موجَّهة ونشرح العمودي المستخدم عند كل سطح فاصل. وأي حلقة تتبع مُعطاة هي معاينة اختيارية، لا عمل أساسي في الأسبوع الثاني. وتتطلب أنماط الدليل الموجي والتشتت وفقد الانحناء والامتصاص فيزياءً أكثر من نموذج الأشعة هذا. وهذا الاستقصاء منفصل عن مكافأة الطور، وهو غير ممتحن وغير شرط مسبق.</p>
</div>

<div dir="rtl" style="text-align:right">
<h2 id="الختام-والتأمل">الختام والتأمل</h2>
<p dir="rtl" style="text-align:right">نشرح نتيجة واحدة بالكلمات، ونذكر وحداتها أو افتراضاتها حين يلزم، ونحدّد فحصًا قد يكشف خطأ. تصف فقرات التمرين والواجب والمكافأة الاختيارية والاستزادة أعلاه الخطوات التالية. لا تصبح الخلفية الاختيارية شرطًا للامتحان.</p>
</div>